In [24]:
#HW1_Part3. Bin Xu

# Instances X: Possible days, each described by the attributes:
# Sky (with possible values Sunny, Cloudy and Rainy),
# AirTemp (with values Warm and Cold),
# Humidity (with values Normal and High),
# Wind (with values Strong and Weak),
# Water (with values Warm and Cool), and
# Forecast (with values Same and Change)
# 
# Hypothesis H：Each hypothesis is described by a conjunction of constraints on the attributes
# Sky, AirTemp, Humidity, Wind, Water, and Forecast. The constraints may be "?"
# (any value is acceptable), "0 (no value is acceptable), or a specific value.
#
#Target concept C: EnjoySport : X->{0,1}
#
#Training examples D: Positive and negative examples of the target function
#
import random
# X is the Instance X
X = [['Sunny','Cloudy','Rainy'],
     ['Warm','Cold'],
     ['Normal','High'],
     ['Strong','Weak'],
     ['Warm','Cool'],
     ['Same','Change']]
# [Sky, AirTemp, Humidity, Wind, Water, Forecast] there are 6 attributes
length_X=len(X)
length_X
# C is Target concept
# D is Training example
# '0' means no value is acceptable
# '? means any value is acceptable
#
#For each positive training instance X
# For each attribute constraint a_i, in H
# If the constraint a_i, is satisfied by X
# Then do nothing
# Else replace a_i, in H by the next more general constraint that is satisfied by X
def findS(D=[]):
    # H is Hypothesis
    H=['0']*length_X
    for constraint in D:
        if constraint[1]:
            for i in range(length_X):
                a_i=constraint[0][i]
                if H[i]=='0':
                    H[i]=a_i
                elif H[i] !='0'and H[i] !=a_i:
                    H[i]='?'
    return H
 
def experiment (C=['?']*length_X):
    S=[]
    #use the Target concept to study the random training examples
    while findS(S) !=C:
        S.append(training_example_generator(C))
    return len(S)

def training_example_generator(C=['?']*length_X):
    D=[]
    #Since the First attribute has 3 elements, so the first attribute has 3 random elements
    D.append(X[0][random.randint(0,2)])
    classification=True
    
    for i in range(1,length_X):
        # The rest attributes have 2 elements, so the rest attributes have 2 random elements
        D.append(X[i][random.randint(0,1)])
        
    for i in range(length_X):
        if C[i] != '?' and C[i] != D[i]:
            classification=False
    return D, classification

def main():
    C=['Sunny','Warm','?','?','?','?']
    number_experiments=500
    H=[]
    for i in range(number_experiments):
        H.append(experiment(C))
    mean_number_of_examples_required=sum(H)/number_experiments
    
    test_verify=[
                (['Sunny', 'Warm', 'Normal', 'Strong', 'Warm', 'Same'], True),
                (['Sunny', 'Warm', 'High', 'Strong', 'Warm', 'Same'], True), 
                (['Rainy', 'Cold', 'High', 'Strong', 'Warm', 'Change'], False),
                (['Sunny', 'Warm', 'High', 'Strong', 'Cool', 'Change'], True),
                ]
    
    verify_hypothsis=findS(test_verify)
    print ('The give targht concept: '+ str(C))
    print ('\n')
    print ('Based on the the sequence of training examples of lecture 3 \n')
    print ('After runing the Find-S algorithm, we have such hypothsis: \n\n'+ str(verify_hypothsis))
    print ('\n')
    print('Experiments times: '+str(len(H)))
    print('Mean number of examples required: ' + str(mean_number_of_examples_required))
            
if __name__ == "__main__":
    main()


The give targht concept: ['Sunny', 'Warm', '?', '?', '?', '?']


Based on the the sequence of training examples of lecture 3 

After runing the Find-S algorithm, we have such hypothsis: 

['Sunny', 'Warm', '?', 'Strong', '?', '?']


Experiments times: 500
Mean number of examples required: 27.526


Frist I used the training examples of lecture 3 to get the hypothsis:
['Sunny', 'Warm', '?', 'Strong', '?', '?']
which is the same as the slide#15 showed

The Target concept is ['Sunny', 'Warm', '?', '?', '?', '?']
After runed the experiment 500 times the mean number of examples required is about 27.5